src/gluonts/dataset/loader_v2.py

```diff
@@ -0,0 +1,189 @@
+# Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License").
+# You may not use this file except in compliance with the License.
+# A copy of the License is located at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# or in the "license" file accompanying this file. This file is distributed
+# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+
+from typing import Iterable, Iterator, List, Callable, Optional
+import itertools
+import random
+from multiprocessing import Process, Queue, Event
+from queue import Empty
+
+import numpy as np
+
+from gluonts.dataset.common import DataBatch, Dataset
+from gluonts.dataset.util import MPWorkerInfo
+
+
+def stack(data, make_array_fn):
+    if isinstance(data[0], np.ndarray):
+        data = make_array_fn(data)
+    elif isinstance(data[0], (list, tuple)):
+        return list(stack(t, make_array_fn) for t in zip(*data))
+    return data
+
+
+def batchify(
+    data: List[dict], make_array_fn: Callable = np.asarray
+) -> DataBatch:
+    return {
+        key: stack(
+            data=[item[key] for item in data], make_array_fn=make_array_fn
+        )
+        for key in data[0].keys()
+    }
+
+
+class CyclicIterable(Iterable):
+    def __init__(self, base_iterable: Iterable) -> None:
+        self.base_iterable = base_iterable
+
+    def __iter__(self):
+        while True:
+            yield from self.base_iterable
+
+
+class PseudoShuffledIterable(Iterable):
+    def __init__(self, base_iterable: Iterable, buffer_length: int):
+        self.base_iterable = base_iterable
+        self.buffer_length = buffer_length
+        self.shuffle_buffer: list = []
+
+    def sample_from_buffer(self):
+        idx = random.randint(0, len(self.shuffle_buffer) - 1)
+        return self.shuffle_buffer.pop(idx)
```

Comment:
I think we should not always use pop when we select an element from the buffer. pop() is much more time consuming than just replacing an index.

Reply:
Good point! Will fix this

Reply:
@yx1215 I actually kept the class we have on master that you implemented: initially I thought that having it in the form of an iterable (as opposed to iterator) would be better, but I think the way you have implemented it is better.